liquidity_pool/src/liquidity.rs

```diff
@@ -54,73 +61,99 @@ pub trait LiquidityModule:
     fn borrow(
         &self,
         initial_caller: Address,
-        collateral_id: TokenIdentifier,
+        collateral_token_id: TokenIdentifier,
         collateral_amount: Self::BigUint,
         deposit_timestamp: u64,
+        ltv: Self::BigUint,
     ) -> SCResult<()> {
-        require!(collateral_amount > 0, "lend amount must be bigger then 0");
-        require!(!initial_caller.is_zero(), "invalid initial caller address");
+        self.require_amount_greater_than_zero(&collateral_amount)?;
+        self.require_non_zero_address(&initial_caller)?;
 
-        let borrows_token = self.borrow_token().get();
-        let asset = self.pool_asset().get();
+        let borrow_token_id = self.borrow_token().get();
+        let pool_token_id = self.pool_asset().get();
+
+        let collateral_data = self.get_token_price_data(&collateral_token_id)?;
+        let pool_asset_data = self.get_token_price_data(&pool_token_id)?;
+
+        let borrowable_amount = self.compute_borrowable_amount(
+            &collateral_amount,
+            &collateral_data.price,
+            &ltv,
+            collateral_data.decimals,
+        );
+
+        let pool_asset_dec_big = Self::BigUint::from(pool_asset_data.decimals as u64);
+        let total_borrowable = (&borrowable_amount / &pool_asset_data.price) / pool_asset_dec_big;
+        let asset_reserve = self.reserves(&pool_token_id).get();
 
-        let mut asset_reserve = self.reserves(&asset).get();
         require!(
-            asset_reserve != Self::BigUint::zero(),
-            "asset reserve is empty"
+            asset_reserve < total_borrowable,
+            "insufficient funds to perform loan"
         );
 
-        let position_id = self.get_nft_hash();
-        let debt_metadata = DebtMetadata {
+        let new_nonce = self.mint_position_tokens(&borrow_token_id, &collateral_amount);
+
+        self.debt_metadata(new_nonce).set(&DebtMetadata {
             timestamp: self.blockchain().get_block_timestamp(),
             collateral_amount: collateral_amount.clone(),
-            collateral_identifier: collateral_id.clone(),
+            collateral_identifier: collateral_token_id,
             collateral_timestamp: deposit_timestamp,
-        };
-
-        self.mint_debt(
-            collateral_amount.clone(),
-            debt_metadata.clone(),
-            position_id.clone(),
-        );
+        });
 
-        let nonce = self
-            .blockchain()
-            .get_current_esdt_nft_nonce(&self.blockchain().get_sc_address(), &borrows_token);
+        self.borrowed_amount()
+            .update(|total| *total += &total_borrowable);
 
-        // send debt position tokens
+        self.reserves(&pool_token_id)
+            .update(|total| *total -= &borrowable_amount);
```

Comment:
`borrowable_amount` is the value in dollars, hope I'm not wrong. Shouldn't it be `total_borrowable` here also? aka the amount of tokens? In whichever case, a better naming will definitely help here, would name them something like: `borrow_amount_in_dollars` and `borrow_amount_in_tokens`

Reply:
done